slamd/discovery/processing/algorithms/discovery_experiment.py

```diff
@@ -81,19 +81,23 @@ def run(self):
         df['Utility'] = df['Utility'].apply(lambda row: round(row, 6))
         df['Novelty'] = df['Novelty'].apply(lambda row: round(row, 6))
 
-        sorted = df.sort_values(by='Utility', ascending=False)
-        # Number the rows from 1 to n (length of the dataframe) to identify them easier on the plots.
-        sorted.insert(loc=0, column='Row number', value=[i for i in range(1, len(sorted) + 1)])
+        sorted = self.preprocess_dataframe_for_output_table(df)
 
         columns_for_plot = self.targets.copy()
         columns_for_plot.extend(['Utility', 'Row number'])
         if len(self.apriori_columns) > 0:
             columns_for_plot.extend(self.apriori_columns)
-
-        candidate_or_target = ['candidate' if row in self.sample_index else 'target' for row in self.features_df.index]
-
         scatter_plot = PlotGenerator.create_target_scatter_plot(sorted[columns_for_plot])
-        tsne_plot = PlotGenerator.create_tsne_input_space_plot(self.features_df, candidate_or_target)
+
+        plot_df = self.features_df.copy()
+        plot_df['is_train_data'] = 'Predicted'
+        plot_df['is_train_data'].iloc[self.sample_index] = 'Labelled'
+        plot_df['Utility'] = -np.inf
+        plot_df['Utility'].iloc[self.prediction_index] = pd.Series(utility_function).values
+        plot_df = plot_df.sort_values(by='Utility', ascending=False)
+        # Number the rows from 1 to n (length of the dataframe) to identify them easier on the plots.
+        plot_df.insert(loc=0, column='Row number', value=[i for i in range(1, len(plot_df) + 1)])
+        tsne_plot = PlotGenerator.create_tsne_input_space_plot(plot_df)
```

Comment:
We could extract a dedicated method for orchestrating the plotting (basically line 86 to 100). This could be called something like _plot_results